cpp/pagerank_module/memgraph_utilities/mg_utility.cpp

```diff
@@ -0,0 +1,286 @@
+#include "mg_procedure.h"
+#include "mg_utility.hpp"
+
+/**
+ * Convert Memgraph list to the edge patterns. Shared nodes are converted into edges.
+ *
+ * @param edge_patterns Label of shared nodes
+ * @return List of patterns
+ */
+std::vector<std::vector<std::string>> ListToPattern(const mgp_list *edge_patterns) {
+    std::vector<std::vector<std::string>> patterns;
+    int size = mgp_list_size(edge_patterns);
+    for (int i = 0; i < size; i++) {
+        std::vector<std::string> pattern;
+
+        const auto *pattern_list = mgp_value_get_list(mgp_list_at(edge_patterns, i));
+        int pattern_size = mgp_list_size(pattern_list);
+        for (int j = 0; j < pattern_size; j++) {
+            std::string label = mgp_value_get_string(mgp_list_at(pattern_list, j));
+            pattern.emplace_back(label);
+        }
+        patterns.emplace_back(pattern);
+    }
+    return patterns;
+}
+
+std::set<int64_t> GetAdjacentVertexId(const mgp_vertex *start_vertex, mgp_memory *memory) {
+    std::set<int64_t> next_vertices;
+    // Iterate on both sides
+    auto *edges_it_out = mgp_vertex_iter_out_edges(start_vertex, memory);
+    auto *edges_it_in = mgp_vertex_iter_in_edges(start_vertex, memory);
+    if (edges_it_out == nullptr || edges_it_in == nullptr) {
+        throw std::runtime_error("Not enough memory!");
+    }
+
+    // For each from, iterate through its edges
+    for (const auto *out_edge = mgp_edges_iterator_get(edges_it_out); out_edge;
+         out_edge = mgp_edges_iterator_next(edges_it_out)) {
+
+        const mgp_vertex *to = mgp_edge_get_to(out_edge);
+        if (start_vertex == to) continue;
+
+        next_vertices.emplace(mgp_vertex_get_id(to).as_int);
+    }
+    // For each from, iterate through its edges
+    for (const auto *in_edge = mgp_edges_iterator_get(edges_it_in); in_edge;
+         in_edge = mgp_edges_iterator_next(edges_it_in)) {
+
+        const mgp_vertex *from = mgp_edge_get_from(in_edge);
+        if (start_vertex == from) continue;
+
+        next_vertices.emplace(mgp_vertex_get_id(from).as_int);
+    }
+    // Destroy edge iterator
+    mgp_edges_iterator_destroy(edges_it_out);
+    mgp_edges_iterator_destroy(edges_it_in);
+
+    return next_vertices;
+}
+
+std::set<int64_t> FindPattern(std::vector<std::string> pattern,
+                              const mgp_vertex *start_vertex,
+                              const mgp_graph *graph,
+                              mgp_memory *memory) {
+
+    int start_index = mgp_vertex_get_id(start_vertex).as_int;
+    std::set<uint64_t> from_vertices;
+    from_vertices.emplace(start_index);
+
+    for (std::string label_name : pattern) {
```

Comment:
`for (const std::string_view label_name : pattern)`

you can then use `label_name.data()` for c-string